### PR TITLE
Set a unique constraint on email table in users #7180

### PR DESCRIPTION
**Pull request by subhamengine**
Link to issue - https://github.com/twentyhq/twenty/issues/7180

File changed - twenty/packages/twenty-server/src/engine/core-modules/user/user.entity.ts

### CODE REVIEW

**Review by bosiraphael — COMMENTED** [review]
Hello @subhamengine ! Thank you for contributing :) I wanted to push a migration script in your PR but it seems that you didn't allow it. Otherwise LGTM.

You can allow it [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) or otherwise just create a file `1726842156293-addUniqueConstraintOnUserEmail` inside `packages/twenty-server/src/database/typeorm/core/migrations` with the following code:

```
import { MigrationInterface, QueryRunner } from 'typeorm';

export class AddUniqueConstraintOnUserEmail1726842156293
  implements MigrationInterface
{
  name = 'AddUniqueConstraintOnUserEmail1726842156293';

  public async up(queryRunner: QueryRunner): Promise<void> {
    await queryRunner.query(
      `ALTER TABLE "core"."user" ADD CONSTRAINT "UQ_e12875dfb3b1d92d7d7c5377e22" UNIQUE ("email")`,
    );
  }

  public async down(queryRunner: QueryRunner): Promise<void> {
    await queryRunner.query(
      `ALTER TABLE "core"."user" DROP CONSTRAINT "UQ_e12875dfb3b1d92d7d7c5377e22"`,
    );
  }
}
```

**Comment by subhamengine** [comment]
@bosiraphael can you check now, is it allowed?
image

**Comment by subhamengine** [comment]
Added the file. Please do check.

**Comment by FelixMalfait** [comment]
FYI @charlesBochet I ran a query to cleanup prod after checking it was inactive workspaces.
It affected 36 users (race condition if you clicked fast then you could create multiple accounts).

I think it's unlikely to be affecting people self-hosting which have much lower volume, but here's the query for reference:

```
WITH RankedUsers AS (
    SELECT *,
           ROW_NUMBER() OVER (PARTITION BY email ORDER BY "createdAt" DESC) as rn
    FROM core.user
)
DELETE FROM core.user
WHERE id IN (
    SELECT id
    FROM RankedUsers
    WHERE rn > 1
);
```

File: packages/twenty-server/src/engine/core-modules/user/user.entity.ts [hunk 1]
```diff
@@ -26,6 +27,7 @@ registerEnumType(OnboardingStatus, {
 });
 
 @Entity({ name: 'user', schema: 'core' })
+@Unique(['email'])
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/core-modules/user/user.entity.ts:30]: I don't think this is needed if you already set it at the column level?

File: packages/twenty-server/src/engine/core-modules/user/user.entity.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ import {
   OneToMany,
   PrimaryGeneratedColumn,
   Relation,
-  UpdateDateColumn,
```
Comment (bosiraphael) [inline packages/twenty-server/src/engine/core-modules/user/user.entity.ts:12]: lint issue ;)

Comment (subhamengine) [reply packages/twenty-server/src/engine/core-modules/user/user.entity.ts:12]: yep

**Review by FelixMalfait — APPROVED** [review]
I then noticed there was a deletedAt so adapted the constraint to take that into account. Thanks for the contribution!

**Pull request merged**